### DISCUSSION TITLE
React 18 & ReactDOM.unmountComponentAtNode() shows warning and does not work

### DISCUSSION THREAD

**Discussion by d-kls** [cypress-io/cypress#26690] [Questions and Help]
Hey i am fairly new to Cypress (& Cypress Component Testing). This is my test case:

```
function TestComponent() {
	// ...some useEffect with unmount functionality
	return <div data-cy="test">...</div>;
}

cy.mount(<TestComponent />).then(() => {
	cy.get("[data-cy='test']")
		.should("be.visible")
		.then(() => {
			// ... doing something
		});

	cy.then(() => ReactDOM.unmountComponentAtNode(getContainerEl()));
	cy.then(() => {
		cy.get("[data-cy='test']")
			.should("not.exist")
			.then(() => {
				// .. do something
		});
	});
});
```

I need to unmount the `TestComponent`, to check some functionality, after having checked it after mounting. I implemented it as referenced in the [docs](https://docs.cypress.io/guides/references/migration-guide#React---unmount-Removed), but this throws a warning inside the console:

![image](https://user-images.githubusercontent.com/68340074/236818170-8899bc90-d7d6-4bff-bb69-661277457c64.png)

while also not unmounting my component:
![image](https://user-images.githubusercontent.com/68340074/236818536-f25951c6-bd56-4a52-b11f-4c3a3f943907.png)

Now before i issue a bug i wanted to check here if perhaps i had missed something about how to unmount the component in React 18 with Cypress correctly.
 
Used versions: 

```
"cypress": "^12.11.0",
"react": "^18.2.0",
"react-dom": "^18.2.0"
```
 
 Thanks in advance.

**Comment by portyu9** [comment]
Hi @d-kls 

This is a React 18 adapter/API mismatch rather than a problem with your component.

For **Cypress 12.11**, `cypress/react` was the React <=17 adapter. Cypress's own source emitted this warning when React 18 was detected and told users to use the React 18 adapter instead.

Separately, React 18 deprecated this legacy API:

```js
ReactDOM.unmountComponentAtNode(...)
```

when using the React 18 root API. A root created with `createRoot()` is supposed to be managed with `root.unmount()`.

You should also avoid manually taking ownership of Cypress's internal React root just to test a component's cleanup.

For your particular test, you can unmount the component **as a child** by rerendering without it:

```jsx
cy.mount(<TestComponent />).then(({ rerender }) => {
  cy.get('[data-cy="test"]').should('be.visible')

  // ...do whatever should happen while mounted

  cy.then(() => rerender(<></>))
})

cy.get('[data-cy="test"]').should('not.exist')
```

That causes React to unmount `TestComponent`, so its `useEffect` cleanup runs, without reaching into Cypress's root implementation.

Another supported behavior since Cypress 11 is that a second mount in the same test removes the previously mounted component:

```jsx
cy.mount(<TestComponent />)

cy.get('[data-cy="test"]').should('be.visible')

// Cypress cleans up the previous mount
cy.mount(<></>)

cy.get('[data-cy="test"]').should('not.exist')
```

For a **current Cypress project**, use the current `cypress/react` adapter; React 18 and React 19 are supported there. `ReactDOM.unmountComponentAtNode()` should not be carried forward — it was deprecated in React 18 and removed completely in React 19.

Please mark this answer as accepted if it helped, thank you.
